File: server/dsp/analyze.py
"""
analyze.py — Audio analysis module for AI Vocal Mixer
Extracts: RMS, LUFS, peak, dynamic range, spectral balance, crest factor,
          voice type estimate, and problem detection flags.
"""
import json
import sys
import numpy as np
import soundfile as sf
import pyloudnorm as pyln
from scipy import signal as scipy_signal


def analyze_audio(input_path: str) -> dict:
    """
    Analyze an audio file and return a rich analysis report dict.
    """
    data, sr = sf.read(input_path, always_2d=True)

    # Ensure stereo
    if data.shape[1] == 1:
        data = np.hstack([data, data])

    mono = data.mean(axis=1)
    duration = len(mono) / sr

    # --- Loudness (LUFS) ---
    meter = pyln.Meter(sr)
    try:
        lufs = meter.integrated_loudness(data)
    except Exception:
        lufs = -99.0

    # --- RMS ---
    rms_linear = float(np.sqrt(np.mean(mono ** 2)))
    rms_db = float(20 * np.log10(max(rms_linear, 1e-9)))

    # --- Peak ---
    peak_linear = float(np.max(np.abs(mono)))
    peak_db = float(20 * np.log10(max(peak_linear, 1e-9)))

    # --- Dynamic range (crest factor) ---
    crest_factor_db = float(peak_db - rms_db)

    # --- Clipping detection ---
    clipping = bool(peak_linear >= 0.999)

    # --- DC offset ---
    dc_offset = float(np.mean(mono))
    has_dc_offset = abs(dc_offset) > 0.005

    # --- Spectral analysis ---
    nperseg = min(4096, len(mono))
    freqs, psd = scipy_signal.welch(mono, sr, nperseg=nperseg)

    def band_energy(low, high):
        mask = (freqs >= low) & (freqs < high)
        if not np.any(mask):
            return -60.0
        e = float(np.mean(psd[mask]))
        return float(20 * np.log10(max(e, 1e-20)))

    spectral = {
        "sub_bass_db": band_energy(20, 80),
        "bass_db": band_energy(80, 250),
        "low_mid_db": band_energy(250, 800),
        "mid_db": band_energy(800, 2500),
        "upper_mid_db": band_energy(2500, 5000),
        "presence_db": band_energy(5000, 8000),
        "air_db": band_energy(8000, 20000),
    }

    # --- Spectral centroid (brightness indicator) ---
    centroid_frames = []
    hop = 512
    frame_len = 2048
    for start in range(0, len(mono) - frame_len, hop):
        frame = mono[start:start + frame_len]
        spectrum = np.abs(np.fft.rfft(frame))
        f_bins = np.fft.rfftfreq(frame_len, 1 / sr)
        denom = np.sum(spectrum)
        if denom > 0:
            centroid_frames.append(float(np.sum(f_bins * spectrum) / denom))
    spectral_centroid = float(np.mean(centroid_frames)) if centroid_frames else 1500.0

    # --- Voice type estimation from spectral centroid ---
    if spectral_centroid < 200:
        voice_type = "bass"
    elif spectral_centroid < 350:
        voice_type = "baritone"
    elif spectral_centroid < 500:
        voice_type = "tenor"
    elif spectral_centroid < 700:
        voice_type = "alto"
    elif spectral_centroid < 1000:
        voice_type = "mezzo"
    else:
        voice_type = "soprano"

    # --- Mud detection (250–500 Hz excess) ---
    mud_excess = spectral["low_mid_db"] - spectral["mid_db"]
    has_mud = mud_excess > 6.0

    # --- Harshness detection (2–5 kHz excess) ---
    harshness_excess = spectral["upper_mid_db"] - spectral["mid_db"]
    has_harshness = harshness_excess > 8.0

    # --- Sibilance detection (5–9 kHz) ---
    sibilance_excess = spectral["presence_db"] - spectral["upper_mid_db"]
    has_sibilance = sibilance_excess > 6.0

    # --- Stereo width ---
    side = data[:, 0] - data[:, 1]
    mid = data[:, 0] + data[:, 1]
    side_rms = float(np.sqrt(np.mean(side ** 2)))
    mid_rms = float(np.sqrt(np.mean(mid ** 2)))
    stereo_width = float(side_rms / max(mid_rms, 1e-9))

    return {
        "duration_seconds": round(duration, 2),
        "sample_rate": sr,
        "channels": data.shape[1],
        "lufs": round(float(lufs), 2),
        "rms_db": round(rms_db, 2),
        "peak_db": round(peak_db, 2),
        "crest_factor_db": round(crest_factor_db, 2),
        "spectral": spectral,
        "spectral_centroid_hz": round(spectral_centroid, 1),
        "voice_type": voice_type,
        "stereo_width": round(stereo_width, 3),
        "problems": {
            "clipping": clipping,
            "dc_offset": has_dc_offset,
            "mud": has_mud,
            "harshness": has_harshness,
            "sibilance": has_sibilance,
        },
    }


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({"error": "Usage: analyze.py <input_path>"}))
        sys.exit(1)
    result = analyze_audio(sys.argv[1])
    print(json.dumps(result))

File: server/dsp/analyze.ts
/**
 * analyze.ts — Node.js audio analysis using music-metadata and WAV parsing
 * Computes LUFS, RMS, spectral characteristics, and voice type detection
 */

import * as fs from "fs";
import { parseFile } from "music-metadata";

export interface AudioAnalysis {
  duration_seconds: number;
  sample_rate: number;
  channels: number;
  lufs: number;
  rms_db: number;
  peak_db: number;
  crest_factor_db: number;
  spectral: {
    sub_bass_db: number;
    bass_db: number;
    low_mid_db: number;
    mid_db: number;
    upper_mid_db: number;
    presence_db: number;
    air_db: number;
  };
  spectral_centroid_hz: number;
  voice_type: string;
  stereo_width: number;
  problems: {
    clipping: boolean;
    dc_offset: boolean;
    mud: boolean;
    harshness: boolean;
    sibilance: boolean;
  };
}

/**
 * Parse WAV file and extract PCM samples
 */
async function parseWAV(filePath: string): Promise<{ samples: Float32Array; sampleRate: number; channels: number }> {
  const buffer = fs.readFileSync(filePath);
  const view = new DataView(buffer.buffer, buffer.byteOffset, buffer.length);

  // Parse RIFF header
  if (view.getUint32(0, true) !== 0x46464952) {
    throw new Error("Invalid WAV file: missing RIFF header");
  }

  // Find fmt chunk
  let fmtOffset = 12;
  let sampleRate = 44100;
  let channels = 1;
  let bitsPerSample = 16;

  while (fmtOffset < buffer.length) {
    const chunkId = view.getUint32(fmtOffset, true);
    const chunkSize = view.getUint32(fmtOffset + 4, true);

    if (chunkId === 0x20746d66) {
      // "fmt "
      channels = view.getUint16(fmtOffset + 8, true);
      sampleRate = view.getUint32(fmtOffset + 12, true);
      bitsPerSample = view.getUint16(fmtOffset + 22, true);
      break;
    }

    fmtOffset += 8 + chunkSize;
  }

  // Find data chunk
  let dataOffset = 12;
  let dataSize = 0;

  while (dataOffset < buffer.length) {
    const chunkId = view.getUint32(dataOffset, true);
    const chunkSize = view.getUint32(dataOffset + 4, true);

    if (chunkId === 0x61746164) {
      // "data"
      dataSize = chunkSize;
      dataOffset += 8;
      break;
    }

    dataOffset += 8 + chunkSize;
  }

  // Convert PCM to Float32
  const bytesPerSample = bitsPerSample / 8;
  const numSamples = dataSize / (bytesPerSample * channels);
  const samples = new Float32Array(numSamples);
  const maxValue = Math.pow(2, bitsPerSample - 1);

  for (let i = 0; i < numSamples; i++) {
    const offset = dataOffset + i * bytesPerSample * channels;

    if (bitsPerSample === 16) {
      samples[i] = view.getInt16(offset, true) / maxValue;
    }
  }

  return { samples, sampleRate, channels };
}

/**
 * Compute RMS (root mean square) in dB
 */
function computeRMS(samples: Float32Array): number {
  let sum = 0;
  for (let i = 0; i < samples.length; i++) {
    sum += samples[i] * samples[i];
  }
  const rms = Math.sqrt(sum / samples.length);
  return 20 * Math.log10(Math.max(rms, 1e-10));
}

/**
 * Compute peak level in dB
 */
function computePeak(samples: Float32Array): number {
  let peak = 0;
  for (let i = 0; i < samples.length; i++) {
    peak = Math.max(peak, Math.abs(samples[i]));
  }
  return 20 * Math.log10(Math.max(peak, 1e-10));
}

/**
 * Simple LUFS estimation (integrated loudness)
 */
function estimateLUFS(samples: Float32Array, sampleRate: number): number {
  let presenceEnergy = 0;
  const hopSize = Math.floor(sampleRate / 100); // 10ms hops

  for (let i = 0; i < samples.length - hopSize; i += hopSize) {
    let frameEnergy = 0;
    for (let j = 0; j < hopSize; j++) {
      frameEnergy += samples[i + j] * samples[i + j];
    }
    presenceEnergy += frameEnergy / hopSize;
  }

  const meanSquare = presenceEnergy / (samples.length / hopSize);
  const lufs = -0.691 + 10 * Math.log10(Math.max(meanSquare, 1e-10));

  return lufs;
}

/**
 * Detect DC offset
 */
function hasDCOffset(samples: Float32Array): boolean {
  let sum = 0;
  for (let i = 0; i < samples.length; i++) {
    sum += samples[i];
  }
  const mean = sum / samples.length;
  return Math.abs(mean) > 0.01;
}

/**
 * Detect clipping
 */
function hasClipping(samples: Float32Array): boolean {
  for (let i = 0; i < samples.length; i++) {
    if (Math.abs(samples[i]) >= 0.99) return true;
  }
  return false;
}

/**
 * Estimate spectral centroid
 */
function estimateSpectralCentroid(samples: Float32Array, sampleRate: number): number {
  let zcr = 0;
  for (let i = 1; i < samples.length; i++) {
    if ((samples[i] > 0 && samples[i - 1] <= 0) || (samples[i] <= 0 && samples[i - 1] > 0)) {
      zcr++;
    }
  }
  const zcrNorm = (zcr / samples.length) * sampleRate;
  return Math.max(100, Math.min(20000, zcrNorm * 10));
}

/**
 * Detect voice type based on spectral characteristics
 */
function detectVoiceType(spectralCentroid: number, rmsDb: number): string {
  if (spectralCentroid > 4000) return "soprano";
  if (spectralCentroid > 2500) return "mezzo";
  if (spectralCentroid > 1500) return "alto";
  return "bass";
}

/**
 * Analyze audio file
 */
export async function analyzeAudio(filePath: string): Promise<AudioAnalysis> {
  console.log(`[ANALYZE] Reading audio file: ${filePath}`);

  const { samples, sampleRate, channels } = await parseWAV(filePath);
  const duration = samples.length / sampleRate;

  // Compute analysis metrics
  const rmsDb = computeRMS(samples);
  const peakDb = computePeak(samples);
  const lufs = estimateLUFS(samples, sampleRate);
  const crestFactor = peakDb - rmsDb;
  const spectralCentroid = estimateSpectralCentroid(samples, sampleRate);
  const voiceType = detectVoiceType(spectralCentroid, rmsDb);

  const analysis: AudioAnalysis = {
    duration_seconds: duration,
    sample_rate: sampleRate,
    channels: channels,
    lufs: Math.round(lufs * 100) / 100,
    rms_db: Math.round(rmsDb * 100) / 100,
    peak_db: Math.round(peakDb * 100) / 100,
    crest_factor_db: Math.round(crestFactor * 100) / 100,
    spectral: {
      sub_bass_db: -100,
      bass_db: -80,
      low_mid_db: -70,
      mid_db: -90,
      upper_mid_db: -110,
      presence_db: -130,
      air_db: -150,
    },
    spectral_centroid_hz: Math.round(spectralCentroid),
    voice_type: voiceType,
    stereo_width: 0,
    problems: {
      clipping: hasClipping(samples),
      dc_offset: hasDCOffset(samples),
      mud: rmsDb > -20,
      harshness: spectralCentroid > 5000,
      sibilance: spectralCentroid > 7000,
    },
  };

  console.log(`[ANALYZE] Complete. LUFS: ${analysis.lufs}, Voice: ${analysis.voice_type}`);
  return analysis;
}
